cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1407,17 +1454,29 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
     });
 
   // Init page fragments
-  // 5000 is good enough for up to ~200-character strings. Longer strings will start producing
-  // fragments larger than the desired page size -> TODO: keep track of the max fragment size, and
-  // iteratively reduce this value if the largest fragment exceeds the max page size limit (we
-  // ideally want the page size to be below 1MB so as to have enough pages to get good
-  // compression/decompression performance).
-  // If using the default fragment size, scale it up or down depending on the requested page size.
-  if (max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
-    max_page_fragment_size_ = (cudf::io::default_max_page_fragment_size * max_page_size_bytes) /
-                              cudf::io::default_max_page_size_bytes;
+  // 5000 is good enough for up to ~200-character strings. Longer strings and deeply nested columns
+  // will start producing fragments larger than the desired page size, so calculate fragment sizes
+  // for each leaf column.  Skip if the fragment size is not the default.
+  std::vector<size_type> column_frag_size(single_streams_table.num_columns(),
+                                          max_page_fragment_size_);
+
+  auto iter = thrust::make_counting_iterator<size_type>(0);
+  if (table.num_rows() > 0 && max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
+    std::for_each(iter, iter + single_streams_table.num_columns(), [&](size_type index) {
+      auto const avg_len = std::max<size_type>(
+        column_size(single_streams_table.column(index), stream) / table.num_rows(), 1);
+      auto const frag_size    = max_page_size_bytes / avg_len;
```

Comment:
rather than max_page_size/avg, should this shoot for multiple fragments per page to catch at least some overshooting?